network/src/testutils/test_node.rs

```diff
@@ -0,0 +1,313 @@
+        let self_metadata = self.conn_metadata(network_id, ConnectionOrigin::Inbound, None);
+        let remote_peer_id = metadata.remote_peer_id;
+        // Tell the other node it's good to send to the connected peer now
+        self.other_inbound_handle(PeerNetworkId::new(network_id, remote_peer_id))
+            .connect(remote_peer_id, network_id, self_metadata);
+        // Then connect us
+        self.node_inbound_handle(network_id)
+            .connect(self_peer_id, network_id, metadata);
+    }
+
+    /// Disconnects a node from another node
+    fn disconnect(&self, network_id: NetworkId, metadata: ConnectionMetadata) {
+        let self_peer_id = self.peer_network_id(network_id).peer_id();
+        let self_metadata = self.conn_metadata(network_id, ConnectionOrigin::Inbound, None);
+        let remote_peer_id = metadata.remote_peer_id;
+        // Tell the other node it's disconnected
+        self.other_inbound_handle(PeerNetworkId::new(network_id, remote_peer_id))
+            .disconnect(remote_peer_id, network_id, self_metadata);
+        // Then disconnect us
+        self.node_inbound_handle(network_id)
+            .disconnect(self_peer_id, network_id, metadata);
+    }
+
+    /// Find a common network between nodes, they're sorted in priority order
+    fn find_common_network(&self, other: &Self) -> Option<NetworkId> {
+        match self.node_type() {
+            NodeType::Validator => match other.node_type() {
+                NodeType::Validator => Some(NetworkId::Validator),
+                NodeType::ValidatorFullNode => Some(NetworkId::Vfn),
+                NodeType::PublicFullNode => None,
+            },
+            NodeType::ValidatorFullNode => match other.node_type() {
+                NodeType::Validator => Some(NetworkId::Vfn),
+                _ => Some(NetworkId::Public),
+            },
+            NodeType::PublicFullNode => match other.node_type() {
+                NodeType::Validator => None,
+                _ => Some(NetworkId::Public),
+            },
+        }
+    }
+
+    /// Build `ConnectionMetadata` for a connection on another node
+    fn conn_metadata(
+        &self,
+        network_id: NetworkId,
+        origin: ConnectionOrigin,
+        maybe_protocols: Option<ProtocolIdSet>,
+    ) -> ConnectionMetadata {
+        let peer_network_id = self.peer_network_id(network_id);
+        let mut metadata = ConnectionMetadata::mock_with_role_and_origin(
+            peer_network_id.peer_id(),
+            self.peer_role(),
+            origin,
+        );
+        if let Some(protocols) = maybe_protocols {
+            metadata.application_protocols = protocols;
+        } else {
+            metadata.application_protocols = ProtocolIdSet::all_known()
+        }
+        metadata
+    }
+}
+
+/// Drops the next queued network message on `Node`'s network (`NetworkId`).  Doesn't propagate
+/// to downstream node
+pub async fn drop_next_network_msg<Node: TestNode>(
```

Comment:
If you rename this to `get_queued_network_msg`, you can reuse it in `send_next_network_msg` below? Technically it doesn't drop the message, but returns it? 😄 The caller has to drop it, right?

Reply:
Yeah, it originally only dropped it, but I changed the signature.  I'll fix this